kernel/src/libs/libUI/textui.rs

```diff
@@ -0,0 +1,1337 @@
+        let buf: TextuiBuf<'_> = TextuiBuf::new(new_buf);
+        return buf;
+    }
+
+    pub fn put_color_in_pixel(&mut self, color: u32, index: usize) {
+        let buf: &mut [u32] = self.0;
+        buf[index] = color;
+    }
+    pub fn get_index_of_next_line(now_index: usize) -> usize {
+        *(TEXTUI_BUF_WIDTH.read()) as usize + now_index
+    }
+    pub fn get_index_by_x_y(x: usize, y: usize) -> usize {
+        *(TEXTUI_BUF_WIDTH.read()) as usize * y + x
+    }
+    pub fn get_start_index_by_lineid_lineindex(lineid: LineId, lineindex: LineIndex) -> usize {
+        //   x 左上角列像素点位置
+        //   y 左上角行像素点位置
+        let index_x: u32 = lineindex.into();
+        let x: u32 = index_x * TEXTUI_CHAR_WIDTH;
+
+        let id_y: u32 = lineid.into();
+        let y: u32 = id_y * TEXTUI_CHAR_HEIGHT;
+
+        TextuiBuf::get_index_by_x_y(x as usize, y as usize)
+    }
+}
+// impl Into<&mut [u32]> for TextuiBuf<'_> {
+//     fn into(&self) -> &mut [u32] {
+//         self.0
+//     }
+// }
+// impl Clone for Buf<'_> {
+//     fn clone(&self) -> Self {
+//         let mut buf = Vec::new();
+//         for i in self.0.iter() {
+//             &buf.push(*i);
+//         }
+//         let b = buf.as_mut_slice();
+//         // 转移对 cloned_data 的所有权到新创建的 Buf 实例中
+//         let cloned = Buf::new(b);
+//         core::mem::forget(b);
+//         return cloned;
+//     }
+// }
+// impl TextuiBuf<'_>{
+//     pub fn new(buf:Buf)->Self{
+//         TextuiBuf(buf)
+//     }
+//     pub fn put_color_in_pixel(&self,color:u32,index:usize){
+//         let buf:&mut [u32]=(*self).into();
+//         buf[index]=color;
+//     }
+//     pub fn get_index_of_next_line(now_index:usize)->usize{
+//         *(BUF_WIDTH.read()) as usize+now_index
+//     }
+//     pub fn get_index_by_x_y(x:usize,y:usize)->usize{
+//         *(BUF_WIDTH.read()) as usize*y+x
+//     }
+//     pub fn get_start_index_by_lineid_lineindex(lineid:LineId,lineindex:LineIndex)->usize{
+//         //   x 左上角列像素点位置
+//         //   y 左上角行像素点位置
+//         let index_x: u32 = lineindex.into();
+//         let x: u32 = index_x * TEXTUI_CHAR_WIDTH;
+//         let id_y: u32 = lineid.into();
+//         let y: u32 = id_y * TEXTUI_CHAR_HEIGHT;
+//         TextuiBuf::get_index_by_x_y(x as usize, y as usize)
+//     }
+// }
+// impl Into<&mut [u32]> for TextuiBuf<'_> {
+//     fn into(self) -> &'static mut [u32] {
+//         self.0.into()
+//     }
+// }
+#[derive(Debug, Copy, Clone, Eq, PartialEq, PartialOrd, Ord, Hash)]
+pub struct Font([u8; 16]);
+impl Font {
+    pub fn get_font(index: usize) -> Font {
+        Self(unsafe { font_ascii[index] })
+    }
+    pub fn is_frcolor(&self, height: usize, width: usize) -> bool {
+        let w = self.0[height];
+        let testbit = 1 << (8 - width);
+        w & testbit != 0
+    }
+}
+
+impl TextuiCharChromatic {
+    fn new() -> Self {
+        TextuiCharChromatic {
+            c: 0,
+            frcolor: FontColor::BLACK,
```

Comment:
Maybe this is a mistake... The front color is black and the background color is black?